### PR TITLE
Add support for EPD 3in7

### PR DESCRIPTION
**Pull request by mangelajo**
Matches implementation from waveshare/e-Paper code

https://github.com/waveshare/e-Paper/blob/master/RaspberryPi_JetsonNano/c/lib/e-Paper/EPD_3in7.c

Co-Authored-By: Miguel Angel Ajo <miguelangel@ajo.es>

### CODE REVIEW

**Comment by mangelajo** [comment]
ok, it's ignoring #![allow(non_camel_case_types)]

I guess I will have to change the enum names.

File: src/epd3in7/command.rs [hunk 1]
```diff
@@ -0,0 +1,89 @@
+//! SPI Commands for the Waveshare 3.7" E-Ink Display
+#![allow(non_camel_case_types)]
```
Comment (caemor) [inline src/epd3in7/command.rs:2]: Please remove the ignore non camel case types and switch the command names to CamelCase type similar as elsewhere in the repository (e.g. https://github.com/caemor/epd-waveshare/blob/main/src/epd2in13bc/command.rs)

Comment (caemor) [reply src/epd3in7/command.rs:2]: I should have read your comment first :see_no_evil: :-D

Comment (mangelajo) [reply src/epd3in7/command.rs]: Ohh, no worries, done :-)

**Comment by mangelajo** [comment]
@caemor is there any driver doing grayscale or shall I define a new color type?

**Comment by mangelajo** [comment]
Ah, there is Gray4! https://docs.rs/embedded-graphics/latest/embedded_graphics/pixelcolor/struct.Gray4.html

**Comment by peckpeck** [comment]
I'd suggest to make the Color a generic parameter and vary it in impl, this way you could factor some of he code

**Comment by mangelajo** [comment]
> I'd suggest to make the Color a generic parameter and vary it in impl, this way you could factor some of he code

Yes, I was thinking about that, but there are a couple of things holding me back:

* The code paths and initialization for both modes (see Gray1/Gray4 functions here [1]) are wildly different, they only seem to share SPI bus access.
* It would special-case this driver in terms of instantiation, I guess we could may be have a default instance in one of the modes -not sure if it's possible, new to rust here-, for other drivers (tri-color vs bw, the drivers are duplicated)

[1] https://github.com/waveshare/e-Paper/blob/master/RaspberryPi_JetsonNano/c/lib/e-Paper/EPD_3in7.c#L171

**Comment by mangelajo** [comment]
Oh, I will go and implement partial updates for this driver first, while I keep rumiating about the Grayscales ;)

**Comment by peckpeck** [comment]
If you use a generic type for color, you can easily separate the code for init with an impl for `Color` and and impl for `Gray4`. But update_frame will still be difficult, maybe smaller trait would help implementing this kind of cases.

**Comment by peckpeck** [comment]
If there is no real improvement besides memory consumption, why have a separate BW initialization ? A BW update_frame on a gray initialized device should work no ?

**Comment by mangelajo** [comment]
> If there is no real improvement besides memory consumption, why have a separate BW initialization ? A BW update_frame on a gray initialized device should work no ?

I really have no numbers to compare yet. I will eventually be working on Gray4 support on a different PR. At that point we can analyze it from both the timing and memory perspective and decide.

But it sounds like a good point, if they were to have the same refresh rates, why keep two code paths :)

**Comment by mangelajo** [comment]
Note: I may need some time to get into Gray4, currently I am working on the application that uses the implementation of this PR.

Hopefully we can merge this if everything looks good, and then treat Gray4 support as an upgrade PR.

**Comment by peckpeck** [comment]
Indeed, since we in 0.x releases we should merge things quickly to allow for easier testing.

File: src/epd3in7/mod.rs [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+
+/// Default Background Color
+pub const DEFAULT_BACKGROUND_COLOR: Color = Color::White;
+
+const IS_BUSY_LOW: bool = false;
+
+/// Display with Fullsize buffer for use with the 2in9 EPD
+#[cfg(feature = "graphics")]
+pub type Display3in7 =
+    crate::graphics::Display<WIDTH, HEIGHT, false, { WIDTH as usize * HEIGHT as usize / 8 }, Color>;
```
Comment (peckpeck) [inline src/epd3in7/mod.rs:35]: As done in https://github.com/caemor/epd-waveshare/pull/128 you should prefer using buffer_len here
` { buffer_len(WIDTH as usize, HEIGHT as usize) }`

Comment (caemor) [inline src/epd3in7/mod.rs:35]:
```suggestion
    crate::graphics::Display<WIDTH, HEIGHT, false,  buffer_len(WIDTH as usize, HEIGHT as usize), Color>;
```

File: src/epd3in7/mod.rs [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+/// Default Background Color
+pub const DEFAULT_BACKGROUND_COLOR: Color = Color::White;
+
+const IS_BUSY_LOW: bool = false;
+
+/// Display with Fullsize buffer for use with the 2in9 EPD
+#[cfg(feature = "graphics")]
+pub type Display3in7 =
+    crate::graphics::Display<WIDTH, HEIGHT, false, buffer_len(WIDTH as usize, HEIGHT as usize), Color>;
```
Comment (caemor) [inline src/epd3in7/mod.rs:35]:
```suggestion
pub type Display3in7 = crate::graphics::Display<
    WIDTH,
    HEIGHT,
    false,
    { buffer_len(WIDTH as usize, HEIGHT as usize) },
    Color,
>;
```

File: src/epd3in7/mod.rs [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+
+/// Height of the display
+pub const HEIGHT: u32 = 480;
+
+/// Default Background Color
+pub const DEFAULT_BACKGROUND_COLOR: Color = Color::White;
+
+const IS_BUSY_LOW: bool = false;
+
+/// Display with Fullsize buffer for use with the 2in9 EPD
```
Comment (mangelajo) [inline src/epd3in7/mod.rs:32]: 2in9->3in7

It doesn't let me make code suggestions .

Comment (peckpeck) [reply src/epd3in7/mod.rs:32]: isn't this your code ?

Comment (mangelajo) [reply src/epd3in7/mod.rs:32]: lol, probably, I thought it was introduced in the changes afterwards.

**Comment by mangelajo** [comment]
> > I am not sure what's going on, I see the errors here:
> 
> You need to rebase your code, the main branch has been updated with new commits that impact your code. Something like `git rebase -i origin/main`

Ack, I will rebase and let's see how it goes, I would have expected the unit testing/etc to pass even if it needed rebase, of course github may complain if it's unable to merge your changes.

File: src/epd3in7/mod.rs (hunk 4 above)
Comment (caemor) [inline src/epd3in7/mod.rs:32]:
```suggestion
/// Display with Fullsize buffer for use with the 3in7 EPD
```

Comment (caemor) [reply src/epd3in7/mod.rs]: I was able to make the code suggestion that you wanted.

**Comment by mangelajo** [comment]
Ok, I rebased and squashed the commits, now at least I get it reproduced locally:

```
error[E0407]: method `is_busy` is not a member of trait `WaveshareDisplay`
   --> src/epd3in7/mod.rs:260:5
    |
260 | /     fn is_busy(&self) -> bool {
261 | |         self.interface.is_busy(IS_BUSY_LOW)
262 | |     }
    | |_____^ not a member of trait `WaveshareDisplay`

error[E0277]: the trait bound `DELAY: _embedded_hal_blocking_delay_DelayUs<u32>` is not satisfied
  --> src/epd3in7/mod.rs:50:37
   |
50 | impl<SPI, CS, BUSY, DC, RST, DELAY> InternalWiAdditions<SPI, CS, BUSY, DC, RST, DELAY>
   |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `_embedded_hal_blocking_delay_DelayUs<u32>` is not implemented for `DELAY`
   |
note: required by a bound in `traits::InternalWiAdditions`
  --> src/traits.rs:36:12
   |
29 | pub(crate) trait InternalWiAdditions<SPI, CS, BUSY, DC, RST, DELAY>
   |                  ------------------- required by a bound in this
...
36 |     DELAY: DelayUs<u32>,
   |            ^^^^^^^^^^^^ required by this bound in `traits::InternalWiAdditions`
help: consider further restricting this bound
   |
58 |     DELAY: DelayMs<u8> + embedded_hal::prelude::_embedded_hal_blocking_delay_DelayUs<u32>,
   |                        ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

error[E0277]: the trait bound `DELAY: _embedded_hal_blocking_delay_DelayUs<u32>` is not satisfied
   --> src/epd3in7/mod.rs:125:37
    |
125 | impl<SPI, CS, BUSY, DC, RST, DELAY> WaveshareDisplay<SPI, CS, BUSY, DC, RST, DELAY>
    |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `_embedded_hal_blocking_delay_DelayUs<u32>` is not implemented for `DELAY`
    |
note: required by a bound in `traits::WaveshareDisplay`
   --> src/traits.rs:146:12
    |
139 | pub trait WaveshareDisplay<SPI, CS, BUSY, DC, RST, DELAY>
    |           ---------------- required by a bound in this
...
146 |     DELAY: DelayUs<u32>,
    |            ^^^^^^^^^^^^ required by this bound in `traits::WaveshareDisplay`
help: consider further restricting this bound
    |
133 |     DELAY: DelayMs<u8> + embedded_hal::prelude::_embedded_hal_blocking_delay_DelayUs<u32>,
    |                        ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
```

I will try to have it fixed by tomorrow.

**Comment by caemor** [comment]
If it helps:
#125 and #131 should be the issues:
- Replace DelayMs<u8> with DelayUs<u32> 
- And replace is_busy with wait_until_idle in the Waveshare Trait Implementation: image

**Comment by mangelajo** [comment]
> If it helps: #125 and #131 should be the issues:
> 
>     * Replace DelayMs with DelayUs
> 
>     * And replace is_busy with wait_until_idle in the Waveshare Trait Implementation: image

Thanks a lot, it helps, yes! #59

**Comment by mangelajo** [comment]
hopefully compile now after rebase :-)

**Review by caemor — APPROVED** [review]

**Comment by mangelajo** [comment]
> Thanks for your great work!

Thank you!! 🙏🏻

**Pull request merged**